modules/ip6/control/router_advertisement.c

```diff
@@ -0,0 +1,264 @@
+	icmp6->code = 0;
+	ra = (struct icmp6_router_advert *)rte_pktmbuf_append(m, sizeof(*ra));
+	ra->cur_hoplim = IP6_DEFAULT_HOP_LIMIT; // Default TTL for this network
+	ra->managed_addr = 0; // DHCPv6 is available
+	ra->other_config = 0; // DNS available, ...
+	ra->lifetime = rte_cpu_to_be_16(5); // Not a default router
+	ra->reachable_time = rte_cpu_to_be_32(0);
+	ra->retrans_timer = rte_cpu_to_be_32(0);
+
+	payload_len = sizeof(*icmp6) + sizeof(*ra);
+
+	if (iface_get_eth_addr(mbuf_data(m)->iface->id, &mac) == 0) {
+		opt = (struct icmp6_opt *)rte_pktmbuf_append(m, sizeof(*opt));
+		opt->type = ICMP6_OPT_SRC_LLADDR;
+		opt->len = ICMP6_OPT_LEN(sizeof(*opt) + sizeof(*lladdr));
+		lladdr = (struct icmp6_opt_lladdr *)rte_pktmbuf_append(m, sizeof(*lladdr));
+		lladdr->mac = mac;
+		payload_len += sizeof(*opt) + sizeof(*lladdr);
+	}
+
+	ip6_set_fields(ip, payload_len, IPPROTO_ICMPV6, &src, &dst);
+	icmp6->cksum = 0;
+	icmp6->cksum = rte_ipv6_udptcp_cksum(ip, icmp6);
+}
+
+static void send_ra_cb(evutil_socket_t, short /*what*/, void *priv) {
+	struct iface *iface = priv;
+	struct rte_mbuf *m;
+
+	struct hoplist *hl = ip6_addr_get_all(iface->id);
+	if (hl == NULL)
+		return;
+	for (unsigned i = 0; i < hl->count; i++) {
+		struct nexthop *nh = hl->nh[i];
+		struct rte_ipv6_addr ip = nh->ipv6;
+		if (nh->family != AF_INET6)
+			continue;
+		if (!rte_ipv6_addr_is_linklocal(&ip))
+			continue;
+
+		if ((m = rte_pktmbuf_alloc(ra_ctx.mp)) == NULL) {
+			LOG(ERR, "rte_pktmbuf_alloc");
+			return;
+		}
+		mbuf_data(m)->iface = iface;
+		build_ra_packet(m, &ip);
+		post_to_stack(ra_output, m);
+	}
+}
+
+static void ra_init(struct event_base *base) {
+	ev_base = base;
+	ra_output = gr_control_input_register_handler("ip6_output", true);
+	ra_ctx.mp = gr_pktmbuf_pool_get(SOCKET_ID_ANY, 512);
+	if (ra_ctx.mp == NULL) {
+		ABORT("gr_pktmbuf_pool_get ENOMEM");
+	}
+}
+
+static void ra_fini(struct event_base * /*ev_base*/) {
+	gr_pktmbuf_pool_release(ra_ctx.mp, 512);
+}
+
+static struct gr_module ra_module = {
+	.name = "ipv6 router advertisement",
+	.init = ra_init,
+	.fini = ra_fini,
+	.fini_prio = 20000,
+};
+
+static struct gr_api_handler ra_set_handler = {
+	.name = "set interface ra",
+	.request_type = GR_IP6_IFACE_RA_SET,
+	.callback = iface_ra_set,
+};
+
+static struct gr_api_handler ra_clear_handler = {
+	.name = "clear interface ra",
+	.request_type = GR_IP6_IFACE_RA_CLEAR,
+	.callback = iface_ra_clear,
+};
+
+static struct gr_api_handler ra_show_handler = {
+	.name = "show interface ra",
+	.request_type = GR_IP6_IFACE_RA_SHOW,
+	.callback = iface_ra_show,
+};
+
+static void iface_event_handler(iface_event_t event, struct iface *iface) {
+	if (event == IFACE_EVENT_POST_ADD) {
+		ra_conf[iface->id].interval = RA_DEFAULT_INTERVAL;
+		ra_conf[iface->id].lifetime = RA_DEFAULT_LIFETIME;
+		ra_conf[iface->id].preference = RA_DEFAULT_PREFERENCE;
+
+		ra_conf[iface->id].timer = event_new(
+			ev_base, -1, EV_PERSIST | EV_FINALIZE, send_ra_cb, iface
+		);
```

Comment:
I never recall this libevent api... Is this adding a 1s timer or do you need to "enable" it afterwards?